src/execution_plan/optimizations/traverse_order.c

```diff
@@ -18,19 +18,23 @@ TRAVERSE_ORDER determineTraverseOrder(const QueryGraph *qg,
                                       size_t expCount) {
 
     if(expCount == 1 || !filterTree) {
-        return TRAVERSE_ORDER_LAST;
+        return TRAVERSE_ORDER_FIRST;
     }
     
     char *destAlias;
     char *srcAlias;
+    bool firstExpLabeled = false;
+    bool lastExpLabeled = false;
     AlgebraicExpression *exp;
-
     TRAVERSE_ORDER order = TRAVERSE_ORDER_FIRST;
+    
     Vector *aliases = FilterTree_CollectAliases(filterTree);
     size_t aliasesCount = Vector_Size(aliases);
 
     // See if there's a filter applied to the first expression.
     exp = exps[0];
+    firstExpLabeled |= (*exp->src_node)->label != NULL;
+    firstExpLabeled |= (*exp->dest_node)->label != NULL;    
```

Comment:
Functionally identical alternate writing:
`bool firstExpLabeled = (*exp->src_node)->label || (*exp->dest_node)->label;`

I don't have a preference on which version is more readable, just putting this out there!